src/vsr/replica.zig

```diff
@@ -5807,11 +5897,64 @@ pub fn ReplicaType(
             const status_before = self.status;
             self.status = .view_change;
 
-            if (self.view == view_new) {
-                assert(status_before == .recovering or status_before == .recovering_head);
-            } else {
+            // Either:
+            // - Transition from normal status.
+            // - Recovering from normal status.
+            // - Retired primary that didn't finish repair.
+            if (status_before == .normal or
+                (status_before == .recovering and self.log_view == self.view) or
+                (status_before == .view_change and self.log_view == self.view))
+            {
+                assert(self.log_view == self.view);
+
+                self.view_headers.array.len = 0;
+
+                const commit_max =
+                    std.math.max(self.commit_max, self.op -| constants.pipeline_prepare_queue_max);
+                var op = self.op + 1;
+                while (op > 0) {
+                    op -= 1;
+
+                    if (self.journal.header_with_op(op)) |header| {
+                        self.view_headers.array.appendAssumeCapacity(header.*);
+
+                        // The DVC headers must connect to the cluster's committed ops.
+                        // We cannot safely go beyond that in all cases:
+                        // - During a prior view-change we might have only accepted a single header
+                        //   from the DVC: "header.op = op_checkpoint_trigger", and then not
+                        //   completed any repair.
+                        // - Similarly, we might have receive a catch-up SV message and only
+                        //   installed a single (checkpoint trigger) hook header.
+                        if (op <= commit_max + 1) break;
+                    } else {
+                        // Skip over the gap.
+                    }
+                }
+                assert(self.view_headers.array.len > 0);
+                assert(self.view_headers.array.len <= constants.pipeline_prepare_queue_max);
+                // We could safely include any additional chained headers.
+                // However, that can make view-change bugs harder to trigger.
```

Comment:
Generate the DVC headers!
(If we can't reuse just them from the previous view).

Note that this is much less aggressive than the old strategy — we include exactly as many headers as we need to to be correct, but no more.